backend/oasst_backend/utils/similarity_functions.py

```diff
@@ -1,17 +1,140 @@
-from typing import List
+import math
 
 import numpy as np
+import scipy.sparse as sp
+import torch
+from sentence_transformers import SentenceTransformer
+from torch import Tensor
+from tqdm import tqdm
 
+ADJACENCY_THRESHOLD = 0.65
 
-def cosine_similarity(a: List[float], b: List[float]):
-    """Compute cosine similarity (dot product of two vectors divided by the product of their norms.)"""
-    norm_a = np.linalg.norm(a)
-    norm_b = np.linalg.norm(b)
-    if norm_a == 0 or norm_b == 0:
-        raise ZeroDivisionError("One of the vectors has a norm of zero.")
-    return np.dot(a, b) / (norm_a * norm_b)
 
+def embed_data(data, key="query", model_name="all-MiniLM-L6-v2", cores=1, gpu=False, batch_size=128):
+    """
+    Embed the sentences/text using the MiniLM language model (which uses mean pooling)
+    """
+    print("Embedding data")
+    model = SentenceTransformer(model_name)
+    print("Model loaded")
 
-def euclidean_distance(a: List[float], b: List[float]):
-    """Compute euclidean distance (norm of the difference of two vectors.)"""
-    return np.linalg.norm(np.subtract(a, b))
+    sentences = data[key].tolist()
+    unique_sentences = data[key].unique()
+    print("Unique sentences", len(unique_sentences))
+
+    if cores == 1:
+        embeddings = model.encode(unique_sentences, show_progress_bar=True, batch_size=batch_size)
+    else:
+        devices = ["cpu"] * cores
+        if gpu:
+            devices = None  # use all CUDA devices
+
+        # Start the multi-process pool on multiple devices
+        print("Multi-process pool starting")
+        pool = model.start_multi_process_pool(devices)
+        print("Multi-process pool started")
+
+        chunk_size = math.ceil(len(unique_sentences) / cores)
+
+        # Compute the embeddings using the multi-process pool
+        embeddings = model.encode_multi_process(unique_sentences, pool, batch_size=batch_size, chunk_size=chunk_size)
+        model.stop_multi_process_pool(pool)
+
+    print("Embeddings computed")
+
+    mapping = {sentence: embedding for sentence, embedding in zip(unique_sentences, embeddings)}
+    embeddings = np.array([mapping[sentence] for sentence in sentences])
+
+    return embeddings
+
+
+def cos_sim(a: Tensor, b: Tensor):
```

Comment:
Is the built-in torch function not suitable here? https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html

Reply:
added built-in torch cosine similarity, I left the old one too incase anyone is interested in the internals. lmk if I should remove